File: mobile-app-frontend/lib/services/wallet/wallet_name_service.dart
import 'package:shared_preferences/shared_preferences.dart';

class WalletNameService {
  static const String _prefsKeyPrefix = 'wallet_name_';

  /// Save a name for a wallet address
  Future<void> saveWalletName(String address, String name) async {
    if (name.trim().isEmpty) {
      await deleteWalletName(address);
      return;
    }
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('$_prefsKeyPrefix${address.toLowerCase()}', name.trim());
  }

  /// Get the name for a wallet address
  Future<String?> getWalletName(String address) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('$_prefsKeyPrefix${address.toLowerCase()}');
  }

  /// Delete the name for a wallet address
  Future<void> deleteWalletName(String address) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('$_prefsKeyPrefix${address.toLowerCase()}');
  }

  /// Get display name (returns name if exists, otherwise returns shortened address)
  Future<String> getDisplayName(String address, {int shortenLength = 6}) async {
    final name = await getWalletName(address);
    if (name != null && name.isNotEmpty) {
      return name;
    }
    return _shortenAddress(address, shortenLength);
  }

  /// Get display name synchronously (for immediate use)
  /// Note: This will return shortened address if name is not yet loaded
  String getDisplayNameSync(String address, {int shortenLength = 6}) {
    return _shortenAddress(address, shortenLength);
  }

  String _shortenAddress(String address, int length) {
    if (address.length < length * 2 + 2) return address;
    return '${address.substring(0, length + 2)}...${address.substring(address.length - length)}';
  }

  /// Get all saved wallet names
  Future<Map<String, String>> getAllWalletNames() async {
    final prefs = await SharedPreferences.getInstance();
    final keys = prefs.getKeys().where((key) => key.startsWith(_prefsKeyPrefix));
    final result = <String, String>{};
    
    for (final key in keys) {
      final address = key.replaceFirst(_prefsKeyPrefix, '');
      final name = prefs.getString(key);
      if (name != null) {
        result[address] = name;
      }
    }
    
    return result;
  }
}

File: ssi-frontend/src/services/wallet/walletNameService.js
const STORAGE_KEY_PREFIX = 'wallet_name_';

const shortenAddress = (address, length = 6) => {
  if (!address || address.length <= length * 2 + 2) {
    return address ?? '';
  }
  return `${address.slice(0, length + 2)}...${address.slice(-length)}`;
};

export class WalletNameService {
  constructor(storage) {
    if (storage) {
      this.storage = storage;
    } else if (typeof window !== 'undefined' && window.localStorage) {
      this.storage = window.localStorage;
    } else {
      this.storage = null;
    }
  }

  _key(address) {
    return `${STORAGE_KEY_PREFIX}${address?.toLowerCase() ?? ''}`;
  }

  async saveWalletName(address, name) {
    if (!this.storage || !address) return;
    const trimmed = name?.trim() ?? '';
    if (!trimmed) {
      await this.deleteWalletName(address);
      return;
    }
    this.storage.setItem(this._key(address), trimmed);
  }

  async getWalletName(address) {
    if (!this.storage || !address) return null;
    return this.storage.getItem(this._key(address));
  }

  async deleteWalletName(address) {
    if (!this.storage || !address) return;
    this.storage.removeItem(this._key(address));
  }

  async getDisplayName(address, shortenLength = 6) {
    const name = await this.getWalletName(address);
    if (name) return name;
    return shortenAddress(address, shortenLength);
  }

  async getAllWalletNames() {
    if (!this.storage) return {};
    const entries = {};
    for (let i = 0; i < this.storage.length; i += 1) {
      const key = this.storage.key(i);
      if (key && key.startsWith(STORAGE_KEY_PREFIX)) {
        const address = key.replace(STORAGE_KEY_PREFIX, '');
        entries[address] = this.storage.getItem(key);
      }
    }
    return entries;
  }
}

export default WalletNameService;

